dc/s2n-quic-dc/src/path/secret/map/state.rs

```diff
@@ -408,13 +409,54 @@ where
     }
 
     fn get_by_addr(&self, peer: &SocketAddr) -> Option<ReadGuard<Arc<Entry>>> {
-        self.peers.get_by_key(peer)
+        self.peers.get_by_key(peer).filter(|_| {
+            // ensure this entry isn't requested to rehandshake
+            !self.requested_handshakes.pin().contains(peer)
+        })
+    }
+
+    fn get_by_addr_tracked(
+        &self,
+        peer: &SocketAddr,
+        handshake: HandshakeKind,
+    ) -> Option<ReadGuard<Arc<Entry>>> {
+        let result = self.peers.get_by_key(peer)?;
+
+        // If this is trying to use a cached handshake but we've got a request to do a handshake, then
+        // force the application to do a new handshake. This is consistent with the `contains` method.
+        if matches!(handshake, HandshakeKind::Cached)
+            && self.requested_handshakes.pin().contains(peer)
+        {
+            return None;
+        }
+
+        self.subscriber().on_path_secret_map_address_cache_accessed(
+            event::builder::PathSecretMapAddressCacheAccessed {
+                peer_address: SocketAddress::from(*peer).into_event(),
+                hit: matches!(handshake, HandshakeKind::Cached),
+            },
+        );
+
+        Some(result)
     }
 
     fn get_by_id(&self, id: &Id) -> Option<ReadGuard<Arc<Entry>>> {
```

Comment:
Maybe should be get_by_id_untracked to make it less likely to accidentally use the shorter form?

Reply:
i like that better! i'll push that real quick